Add generics support for queries and lookups

enyo opened the pull request:
This is a breaking change that supports generics which makes it easier to be used with dart 2.0

```dart
// Before
db.query(Person).run(); // -> Stream<Model>
// After
db.query<Person>().run(); // -> Stream<Person>

// Before
db.lookup([key]); // -> Future<List<Model>>
// After
db.lookup<Person>([key]); // -> Future<List<Person>>
```

Closes #54

mkustermann commented on test/datastore/error_matchers.dart:
```diff
@@ -9,39 +9,32 @@ import 'dart:io';
-class _ApplicationError extends TypeMatcher {
-  const _ApplicationError() : super("ApplicationError");
-  bool matches(item, Map matchState) => item is ApplicationError;
+class _ApplicationError extends TypeMatcher<ApplicationError> {
```
Actually, I think we can get rid of the classes entirely now:

```dart
const isApplicationError = const TypeMatcher<ApplicationError>()
```

enyo replied:
Agreed.

enyo replied:
Implemented.

mkustermann commented on lib/src/db/db.dart:
```diff
@@ -399,12 +398,12 @@ Future _commitHelper(DatastoreDB db,
       .lookup(entityKeys, transaction: datastoreTransaction)
       .then((List<ds.Entity> entities) {
-    return entities.map(db.modelDB.fromDatastoreEntity).toList();
+    return entities.map(db.modelDB.fromDatastoreEntity).cast<T>().toList();
```
Since `fromDatastoreEntity` is now `fromDatastoreEntity<T>` we should be able to avoid the extra `.cast<T>()` here. Something like this should work:

```dart
return entities.map<T>(db.modelDB.fromDatastoreEntity).toList()
```

This uses a new language feature in Dart 2.0 which is a "partial instantiation". Implicitly it should do this:


```dart
var closure = db.modelDB.fromDatastoreEntity;
var closureWithType = closure<T>;
return entities.map<T>(closure).toList()
```

Disclaimer: Have not tested it.

enyo replied:
Nice! Added the change.

enyo commented:
Great. I’ll make the changes and additional cleanups on Monday.

enyo commented:
> I think before merging this PR, we should ensure that the rest of the code is Dart 2.0 compliant (no upper case UTF8/...). If there's something missing we would need to fix those first and publish a new version before merging this PR.

I haven't seen any of these, nor were there any warnings.

But I now added `implicit-casts: false` to the strong-mode settings, and there are lots of issues. I will take care of them.

enyo commented:
@mkustermann I've now added `implicit-casts: false` and fixed all >200 errors. They have been added as two separate commits, one for the library itself, and one for all the tests.

enyo commented:
Oh,... I just realised that you wanted to merge the fixes for dart 2.0 first separately before making the changes. Do you still want to do this?
In that case, I would need to split this merge request. Not too big of an issue.

enyo commented:
@mkustermann do you want me to split the PR into two separate ones? One that is a simple dart 2.0 upgrade, and the other one with the breaking generics changes?

mkustermann commented:
@enyo Apologies for the delay, I will try to take a look at this soon.

mkustermann commented on lib/src/db/db.dart:
```diff
@@ -228,15 +229,15 @@ class Query {
       return _db.datastore
           .query(query, transaction: _transaction, partition: partition);
-    }).stream.map(_db.modelDB.fromDatastoreEntity);
+    }).stream.map((entity) => _db.modelDB.fromDatastoreEntity<T>(entity));
```
I believe it should be possible to use:
```dart
}).stream.map<T>(_db.modelDB.fromDatastoreEntity);
```
to avoid the additional closure.

mkustermann commented on lib/src/db/model_db_impl.dart:
```diff
@@ -412,7 +413,7 @@ class _ModelDescription<T extends Model> {
   }
 
-  Model decodeEntity(ModelDBImpl db, Key key, ds.Entity entity) {
+  T decodeEntity<T extends Model>(ModelDBImpl db, Key key, ds.Entity entity) {
```
Maybe name the type parameter here `H`, since it shadows `_ModelDescription.T`.

mkustermann commented on lib/src/db/model_db_impl.dart:
```diff
@@ -513,7 +514,8 @@ class _ExpandoModelDescription extends _ModelDescription<ExpandoModel> {
     });
-    return model;
+    // TODO: check if there is a more elegant solution than this
+    return model as T;
```
Would it address this TODO if we changed
```dart
T decodeEntity<T extends Model>(ModelDBImpl db, Key key, ds.Entity entity) {
  ExpandoModel model = super.decodeEntity(db, key, entity);
  ...
  // TODO: check if there is a more elegant solution than this
  return model as T;
}
```
to
```dart
T decodeEntity<T extends ExpandoModel>(ModelDBImpl db, Key key, ds.Entity entity) {
  T model = super.decodeEntity<T>(db, key, entity);
  ...
  return mode;
}
```
?

enyo replied:
The problem is, that using `T` instead of `ExpandoModel` means that there is no `additionalProperties` attribute, that is used just below.

mkustermann replied:
When using `T` and having a type-bound of `T extends ExpandoModel` then we should have all the properties of `ExpandoModel`.

enyo replied:
Oh sorry. I didn't see that you meant extending `ExpandoModel`. 

Well that would work in that specific case, but in every place this function is used we then get a type error because `T extends Model` is not `T extends ExpandoModel`

mkustermann replied:
Aha. Okay, so that doesn't work. Makes sense.

mkustermann commented:
Hope it's okay that I squash it into 1 commit :)

enyo commented:
Sure :)
Have you given the generics Key problematic some thought? I have an implementation ready, with generic keys that we already started using in our app, and it works quite well, except for parents keys.

kevmoo commented on .gitignore:
```diff
@@ -3,3 +3,4 @@ pubspec.lock
 packages
 .pub
 .packages
+.idea
```
Please don't add IDE-specific ignores! See https://www.dartlang.org/guides/libraries/private-files

enyo replied:
Oh sorry! Didn't know that was a guideline.

enyo replied:
Mh, on the page you linked there is no mention of not adding IDE-specific ignores (only that one shouldn't actually commit IDE-specific files).

mkustermann replied:
(I think Kevin was referring to "To avoid committing files that are specific to your personal workflow or setup, consider using a global ignore file (for example, .gitignore_global)".)

Though I don't mind having workflow specific files in `.gitignore`. It's better than someone accidentally committing editor specific things into the repository.
